source/addonHandler/__init__.py

```diff
@@ -334,19 +328,27 @@ def completeRemove(self,runUninstallTask=True):
 		shutil.rmtree(tempPath,ignore_errors=True)
 		if os.path.exists(tempPath):
 			log.error("Error removing addon directory %s, deferring until next NVDA restart"%self.path)
-		# clean up the addons state
-		log.debug("removing addon {} from _disabledAddons".format(self.name))
+		# clean up the addons state. If an addon with the same name is installed, it should not be automatically
+		# disabled / blocked.
+		log.debug("removing addon {} from _disabledAddons/_blockedAddons".format(self.name))
 		_disabledAddons.discard(self.name)
+		_blockedAddons.discard(self.name)
 		saveState()
 
 	def addToPackagePath(self, package):

```

Comment:
> A first call of this function would also yield add-ons that ought to be blocked.

I don't think this is true. Not that in `_getAvailableAddonsFromPath` the `addon.name` is added to `_blockedAddons` when the addon is incompatible.  These addons then return `False` from `isRunning()` because they are blocked. `isRunning()` is the filter function used with `getRunningAddons`

Reply:
Ah yes, you're right. Still, it might be good to know what you think of my proposed alternative.

Reply:
I'm looking into your suggestion now.

Since it may be handy to refer back to, my understanding of how addons are enabled is thus:

Various modules which need to collect and call methods on the addons or other global plugins (to make them "run") do by getting the `globalPluginHandler.runningPlugins`. This `runningPlugins` set is created in `globalPluginHandler.initialize` by iterating over `listPlugins()` which seems to be iterating on modules in `globalPlugins.__path__`. I think this is initially populated by `config.addConfigDirsToPythonPackagePath(globalPlugins)` which iterates over all the "runningAddons" and calls `addon.addToPackagePath`

This seems to be why excluding an addon (eg by name), from `addon.addToPackagePath` means it will not be run.

I don't think this PR is a good place to address this, but It would be great if there was a more direct and obvious mechanism for this. For instance a function "loadAndStartAddons". It's still not clear to me that this is all that is required for the addons to run, and perhaps there is some other special treatment.

Reply:
Yes, I think we both agree that this could be tidied up, and I think the approach you suggested would be a good place to start. It would certainly make it much more explicit about what is happening. One concern is that there are modules other than "appModules, brailleDisplayDrivers, synthDrivers, globalPlugins" that addons rely on being added to the package path.

Perhaps "run" is poor word to use, but it's suggested by functions like `getRunningAddons` and `addon.isRunning`, perhaps "registered" is better way of referring to this action, or as you have suggested (perhaps accidentally) "activated". I almost want to call it "installed", except for the confusion that creates with the copying of files".

Reply:
>  One concern is that there are modules other than "appModules, brailleDisplayDrivers, synthDrivers, globalPlugins" that addons rely on being added to the package path.

Reading this back, I think I should clarify what I meant. I am concerned that addons may exist that rely on other modules being added to the package paths. If we restrict this to only look for a specific set of modules, we may break addons in a new way.

I'm going to mark this conversation as resolved, and try to push forward with getting this PR merged.